rclrs/src/wait/guard_condition.rs

```diff
@@ -0,0 +1,149 @@
+use crate::rcl_bindings::*;
+use crate::{Context, RclrsError, ToResult};
+
+use std::sync::{atomic::AtomicBool, Arc, Mutex, MutexGuard};
+
+/// A struct for encapsulating a guard condition - a waitable trigger
+///
+/// # Example
+/// ```
+/// # use rclrs::{Context, GuardCondition, WaitSet, RclrsError};
+/// # use std::sync::{Arc, atomic::Ordering};
+///
+/// let context = Context::new([])?;
+///
+/// let atomic_bool = Arc::new(std::sync::atomic::AtomicBool::new(false));
+/// let atomic_bool_for_closure = Arc::clone(&atomic_bool);
+///
+/// let gc = Arc::new(GuardCondition::new(
+///     &context,
+///     Some(Box::new(move || {
+///         atomic_bool_for_closure.store(true, Ordering::Relaxed);
+///     })),
+/// ));
+///
+/// let mut ws = WaitSet::new(0, 1, 0, 0, 0, 0, &context)?;
+/// ws.add_guard_condition(Arc::clone(&gc))?;
+///
+/// // Trigger the guard condition, firing the callback and waking any wait sets being waited on
+/// gc.trigger()?;
+///
+/// // The wait call will now immediately return
+/// ws.wait(Some(std::time::Duration::from_millis(10)))?;
+///
+/// # Ok::<(), RclrsError>(())
+/// ```
+pub struct GuardCondition {
+    /// The rcl_guard_condition_t that this struct encapsulates.
+    rcl_guard_condition: Arc<Mutex<rcl_guard_condition_t>>,
+    /// An optional callback to call when this guard condition is triggered.
+    callback: Option<Box<dyn Fn() + Send>>,
+    /// A flag to indicate if this guard condition has already been assigned to a wait set.
+    pub(crate) in_use_by_wait_set: Arc<AtomicBool>,
+}
+
+impl Drop for GuardCondition {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: No precondition for this function (besides passing in a valid guard condition)
+            rcl_guard_condition_fini(&mut *self.rcl_guard_condition.lock().unwrap());
+        }
+    }
+}
+
+// SAFETY: rcl_guard_condition is the only member that doesn't implement Send, and it is designed to be accessed from other threads
+unsafe impl Send for GuardCondition {}
+
+// SAFETY: The mutexes and atomic members ensure synchronized access to members, and the callback is reentrant
+unsafe impl Sync for GuardCondition {}
+
+impl GuardCondition {
+    /// Creates a new guard condition
+    pub fn new(context: &Context, callback: Option<Box<dyn Fn() + Send>>) -> Self {
```

Comment:
Another usability idea: We can return `Arc<Self>` here. You'll never want to use one without `Arc`.